contracts/rbn-staking/IncentivizedVotingLockup.sol

```diff
@@ -0,0 +1,805 @@
+
+  /** Shared Events */
+  event Deposit(
+    address indexed provider,
+    uint256 value,
+    uint256 locktime,
+    LockAction indexed action,
+    uint256 ts
+  );
+  event Withdraw(address indexed provider, uint256 value, uint256 ts);
+
+  // RBN Redeemer contract
+  IRBNRedeemer public rbnRedeemer;
+
+  /** Shared Globals */
+  IERC20 public stakingToken;
+  uint256 private constant WEEK = 7 days;
+  uint256 public constant MAXTIME = 4 * 365 days; // 4 years
+
+  /** Lockup */
+  uint256 public globalEpoch;
+  uint256 public totalLocked;
+  Point[] public pointHistory;
+  mapping(address => Point[]) public userPointHistory;
+  mapping(address => uint256) public userPointEpoch;
+  mapping(uint256 => int128) public slopeChanges;
+  mapping(address => LockedBalance) public locked;
+
+  // Voting token - Checkpointed view only ERC20
+  string public name;
+  string public symbol;
+  uint256 public decimals = 18;
+
+  /** Structs */
+  struct Point {
+    int128 bias;
+    int128 slope;
+    uint256 ts;
+    uint256 blk;
+  }
+
+  struct LockedBalance {
+    int128 amount;
+    uint256 end;
+  }
+
+  enum LockAction {CREATE_LOCK, INCREASE_LOCK_AMOUNT, INCREASE_LOCK_TIME}
+
+  constructor(
+    address _stakingToken,
+    address _rbnRedeemer,
+    string memory _name,
+    string memory _symbol
+  ) {
+    require(_stakingToken != address(0), "!_stakingToken");
+    require(_rbnRedeemer != address(0), "!_rbnRedeemer");
+
+    stakingToken = IERC20(_stakingToken);
+    Point memory init =
+      Point({
+        bias: int128(0),
+        slope: int128(0),
+        ts: block.timestamp,
+        blk: block.number
+      });
+    pointHistory.push(init);
+
+    decimals = IERC20Detailed(_stakingToken).decimals();
+    require(decimals <= 18, "Cannot have more than 18 decimals");
+
+    rbnRedeemer = _rbnRedeemer;
+    name = _name;
+    symbol = _symbol;
+  }
+
+  /**
+   * @dev Validates that the user has an expired lock && they still have capacity to earn
+   * @param _addr User address to check
+   */
+  modifier lockupIsOver(address _addr) {
+    LockedBalance memory userLock = locked[_addr];
+    require(
+      userLock.amount > 0 && block.timestamp >= userLock.end,
+      "Users lock didn't expire"
+    );
+    require(staticBalanceOf(_addr) > 0, "User must have existing bias");
+    _;
+  }
+
+  /**
+   * @dev Validates that the tx sender is rbnRedeemer contract
+   */
+  modifier onlyRBNRedeemer() {
+    require(msg.sender == rbnRedeemer, "Must be rbn redeemer contract");
+    _;
+  }
+
+  /***************************************
+                LOCKUP - GETTERS
+    ****************************************/
+
+  /**
+   * @dev Redeems rbn to redeemer contract in case criterium met (i.e smart contract hack, vaults get rekt)
+   * @param _amount amount to withdraw to redeemer contract
+   */
+  function redeemRBN(uint256 _amount) external onlyRBNRedeemer {
+    stakingToken.safeTransfer(rbnRedeemer.admin(), _amount);
```

Comment:
One improvement we can make here is to create a new `rbnRedeemer.recipient` variable to differentiate the admin and the RBN recipient

Makes it so that we can make them 2 different things

Reply:
Actually - better yet, the VotingLockup contract doesn't need to have knowledge of the `recipient` or `admin`, it can just transfer to the rbnRedeemer contract directly and the contract does whatever it needs. like this:

```solidity
stakingToken.safeTransfer(rbnRedeemer, _amount);
```

Reply:
👍